Catalyst3/university-microservices
# Added Resilience4j circuit breaker with fallback and actuator configu…

Title: Add Resilience4j Circuit Breaker, Fallback Handling, and Actuator Configuration

Description:

This PR implements the following key features using Resilience4j:

- Circuit Breaker Configuration:
  - Configured sliding window size, failure threshold, and duration in the open state.
  - Set the number of calls allowed in the open state before transitioning to the half-open state.
  - Implemented a fallback method to handle failures when the circuit is open.

- AOP Separation:
  - Moved circuit breaker logic to a separate class since AOP does not work within the same class, ensuring clean separation of concerns.

- Actuator Configuration:
  - Added actuator endpoints to monitor the health and status of the circuit breaker.

Testing Steps:
1. Service Setup:
   - Ran Eureka Server, API Gateway, and Student Service while keeping Address Service down to simulate failure scenarios.

2. Testing URLs:
   - Service Call: http://localhost:9090/student-service/api/student/getById/1
     - Tested the Student Service, which makes an internal call to the Address Service. This call triggers the circuit breaker and invokes the fallback when Address Service is down.
   - Health Check: http://localhost:9090/student-service/actuator/health
     - Verified the health status of the Student Service.

Next Steps:
- Perform further testing under various load conditions.
- Monitor and fine-tune the circuit breaker settings using actuator metrics.
